Gradle + quarkus-container-image-jib: AccessDeniedException on second run on Windows

**Describe the bug**
I try to build a container image with the `quarkus-container-image-jib` extension in my Quarkus (gradle, Kotlin) project on my Windows 10 PC. The first run works (`./gradlew assemble -Dquarkus.container-image.build=true`). On the second and any further runs l I always get `AccessDeniedException`s. When I stop the gradle deamons (`./gradlew --stop`) or kill them via the TaskManager, it works again for one run but then I end up with the same issue.

Full exception:
```
17:15:27: Executing task 'assemble -Dquarkus.container-image.build=true'...


> Task :quarkusGenerateCode
preparing quarkus application
kotlin scripting plugin: root project 'code-with-quarkus'.compileQuarkus-generated-sourcesKotlin - configuration not found: quarkus-generated-sourcesKotlinScriptDef, the plugin is probably applied by a mistake

> Task :quarkusGenerateCodeTests
preparing quarkus application
kotlin scripting plugin: root project 'code-with-quarkus'.compileQuarkus-test-generated-sourcesKotlin - configuration not found: quarkus-test-generated-sourcesKotlinScriptDef, the plugin is probably applied by a mistake

> Task :compileKotlin
> Task :compileJava NO-SOURCE
> Task :processResources UP-TO-DATE
> Task :classes UP-TO-DATE
> Task :inspectClassesForKotlinIC
> Task :jar

> Task :quarkusBuild FAILED
building quarkus jar
7 actionable tasks: 6 executed, 1 up-to-date

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':quarkusBuild'.
> io.quarkus.builder.BuildException: Build failure: Build failed due to errors
  	[error]: Build step io.quarkus.deployment.pkg.steps.JarResultBuildStep#buildRunnerJar threw an exception: java.lang.IllegalStateException: java.nio.file.AccessDeniedException: C:\development\quarkus-debugging\bug-reports\2020-09-04_jib-problems\build\lib
  	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:941)
  	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
  	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
  	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2046)
  	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1578)
  	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
  	at java.base/java.lang.Thread.run(Thread.java:830)
  	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
  Caused by: java.nio.file.AccessDeniedException: C:\development\quarkus-debugging\bug-reports\2020-09-04_jib-problems\build\lib
  	at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:89)
  	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
  	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
  	at java.base/sun.nio.fs.WindowsFileSystemProvider.createDirectory(WindowsFileSystemProvider.java:506)
  	at java.base/java.nio.file.Files.createDirectory(Files.java:693)
  	at java.base/java.nio.file.Files.createAndCheckIsDirectory(Files.java:800)
  	at java.base/java.nio.file.Files.createDirectories(Files.java:786)
  	at io.quarkus.deployment.pkg.steps.JarResultBuildStep.buildLegacyThinJar(JarResultBuildStep.java:418)
  	at io.quarkus.deployment.pkg.steps.JarResultBuildStep.buildRunnerJar(JarResultBuildStep.java:192)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
  	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:932)
  	... 7 more


* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 4s
17:15:32: Task execution finished 'assemble -Dquarkus.container-image.build=true'.
```


**To Reproduce**
Steps to reproduce the behavior:
1. I assume you need a Windows machine (I have Windows 10)
2. Download reproducer project and unzip it
[2020-09-04_jib-problems.zip](https://github.com/quarkusio/quarkus/files/5175659/2020-09-04_jib-problems.zip)
3. Run `./gradlew assemble -Dquarkus.container-image.build=true`
4. Change something in the code (e.g. in `ExampleResource`)
5. Run `./gradlew assemble -Dquarkus.container-image.build=true` again (now it should crash with the `AccessDeniedException`)

**Environment (please complete the following information):**
 - Windows 10
- Gradle
- Kotlin
- OpenJDK 11

--- a/core/deployment/src/main/java/io/quarkus/deployment/mutability/DevModeTask.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/mutability/DevModeTask.java
@@ -129,8 +129,7 @@ public class DevModeTask {
                 //not all local projects are dependencies
                 continue;
             }
-            IoUtils.recursiveDelete(moduleClasses);
-            Files.createDirectories(moduleClasses);
+            IoUtils.recursiveDeleteAndThenCreate(moduleClasses);
             for (Path p : dep.getPaths()) {
                 if (Files.isDirectory(p)) {
                     Path moduleTarget = moduleClasses;

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/AppCDSBuildStep.java
@@ -38,8 +38,7 @@ public class AppCDSBuildStep {
     public void requested(OutputTargetBuildItem outputTarget, BuildProducer<AppCDSRequestedBuildItem> producer)
             throws IOException {
         Path appCDSDir = outputTarget.getOutputDirectory().resolve("appcds");
-        IoUtils.recursiveDelete(appCDSDir);
-        Files.createDirectories(appCDSDir);
+        IoUtils.recursiveDeleteAndThenCreate(appCDSDir);
 
         producer.produce(new AppCDSRequestedBuildItem(outputTarget.getOutputDirectory().resolve("appcds")));
     }

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java
@@ -414,8 +414,7 @@ public class JarResultBuildStep {
                 .resolve(outputTargetBuildItem.getBaseName() + packageConfig.runnerSuffix + ".jar");
         Path libDir = outputTargetBuildItem.getOutputDirectory().resolve("lib");
         Files.deleteIfExists(runnerJar);
-        IoUtils.recursiveDelete(libDir);
-        Files.createDirectories(libDir);
+        IoUtils.recursiveDeleteAndThenCreate(libDir);
 
         try (FileSystem runnerZipFs = ZipUtils.newZip(runnerJar)) {
 
@@ -464,8 +463,7 @@ public class JarResultBuildStep {
             userProviders = buildDir.resolve(packageConfig.userProvidersDirectory.get());
         }
         if (!rebuild) {
-            IoUtils.recursiveDelete(buildDir);
-            Files.createDirectories(buildDir);
+            IoUtils.recursiveDeleteAndThenCreate(buildDir);
             Files.createDirectories(mainLib);
             Files.createDirectories(baseLib);
             Files.createDirectories(appDir);
@@ -477,8 +475,7 @@ public class JarResultBuildStep {
                 Files.createFile(userProviders.resolve(".keep"));
             }
         } else {
-            IoUtils.recursiveDelete(quarkus);
-            Files.createDirectories(quarkus);
+            IoUtils.recursiveDeleteAndThenCreate(quarkus);
         }
         Map<AppArtifactKey, List<Path>> copiedArtifacts = new HashMap<>();
 
@@ -712,8 +709,7 @@ public class JarResultBuildStep {
             List<UberJarRequiredBuildItem> uberJarRequired) throws Exception {
         Path targetDirectory = outputTargetBuildItem.getOutputDirectory()
                 .resolve(outputTargetBuildItem.getBaseName() + "-native-image-source-jar");
-        IoUtils.recursiveDelete(targetDirectory);
-        Files.createDirectories(targetDirectory);
+        IoUtils.recursiveDeleteAndThenCreate(targetDirectory);
 
         List<GeneratedClassBuildItem> allClasses = new ArrayList<>(generatedClasses);
         allClasses.addAll(nativeImageResources.stream()

--- a/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/bootstrap/util/IoUtils.java
+++ b/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/bootstrap/util/IoUtils.java
@@ -6,6 +6,7 @@ import java.io.InputStream;
 import java.io.OutputStream;
 import java.io.StringWriter;
 import java.nio.charset.StandardCharsets;
+import java.nio.file.AccessDeniedException;
 import java.nio.file.FileAlreadyExistsException;
 import java.nio.file.FileVisitOption;
 import java.nio.file.FileVisitResult;
@@ -86,6 +87,36 @@ public class IoUtils {
                 }
             });
         } catch (IOException e) {
+        }
+    }
+
+    public static void recursiveDeleteAndThenCreate(Path root) throws IOException {
+        recursiveDelete(root);
+        if (!PropertyUtils.isWindows()) {
+            Files.createDirectories(root);
+            return;
+        }
+
+        // in Windows, recreating a directory right after deleting it can be problematic, see https://bugs.openjdk.java.net/browse/JDK-8029608
+        // so we just try the create a operation a few times and hope it works
+
+        final int maxTries = 3;
+        int i = 0;
+        while (true) {
+            try {
+                Files.createDirectories(root);
+                return;
+            } catch (AccessDeniedException e) {
+                if (++i == maxTries) {
+                    throw e;
+                }
+
+                try {
+                    Thread.sleep(100);
+                } catch (InterruptedException ignored) {
+
+                }
+            }
         }
     }
 

--- a/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/bootstrap/util/PropertyUtils.java
+++ b/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/bootstrap/util/PropertyUtils.java
@@ -21,7 +21,7 @@ public class PropertyUtils {
     }
 
     public static boolean isWindows() {
-        return getProperty(OS_NAME).toLowerCase(Locale.ENGLISH).indexOf(WINDOWS) >= 0;
+        return getProperty(OS_NAME).toLowerCase(Locale.ENGLISH).contains(WINDOWS);
     }
 
     public static String getUserHome() {

